File: eng/docs/agent_skills/csharp_typing/examples/integration_tests/GuardBehaviorTest.cs
// ============================================================
// Integration test: Guard behavior
// Family: guard_behavior
// Level: cmp_int (component integration — guard + provider)
// ============================================================
//
// Verifies that the guard correctly rejects duplicate in-flight
// requests and that the inner provider is not called twice.
// Uses threads to simulate concurrent calls.

using FluentAssertions;
using Xunit;

// Adjust these usings to your real namespace.

namespace Acme.Orders.Tests.Integration;

// ============================================================
// Domain types (self-contained example)
// ============================================================

internal sealed record OrderAccepted(string OrderId, string Status);

internal sealed record SubmitOrderGuardConfig(bool SingleInflight = true);

internal sealed class SubmitOrderGuardState
{
    private readonly HashSet<string> _inflightIds = [];
    private readonly object _lock = new();

    public bool TryAdd(string orderId)
    {
        lock (_lock) { return _inflightIds.Add(orderId); }
    }

    public void Remove(string orderId)
    {
        lock (_lock) { _inflightIds.Remove(orderId); }
    }
}

// ============================================================
// Event capture
// ============================================================

internal sealed class EventCapture
{
    private readonly List<Dictionary<string, object>> _events = [];
    private readonly object _lock = new();

    public void Emit(Dictionary<string, object> ev)
    {
        lock (_lock) { _events.Add(ev); }
    }

    public List<Dictionary<string, object>> Events
    {
        get { lock (_lock) { return [.. _events]; } }
    }
}

// ============================================================
// Blocking provider (simulates slow inner call)
// ============================================================

internal sealed class BlockingOrderProvider
{
    private readonly ManualResetEventSlim _entered = new(false);
    private readonly ManualResetEventSlim _release = new(false);

    public List<string> Calls { get; } = [];
    public ManualResetEventSlim Entered => _entered;
    public ManualResetEventSlim Release => _release;

    public OrderAccepted SubmitOrder(string orderId)
    {
        Calls.Add(orderId);
        _entered.Set();
        _release.Wait(TimeSpan.FromSeconds(1));
        return new OrderAccepted(orderId, "accepted");
    }
}

// ============================================================
// Guard under test
// ============================================================

internal sealed class GuardedOrderPort
{
    private readonly BlockingOrderProvider _inner;
    private readonly SubmitOrderGuardConfig _config;
    private readonly EventCapture _obs;
    private readonly SubmitOrderGuardState _state = new();

    public GuardedOrderPort(
        BlockingOrderProvider inner,
        SubmitOrderGuardConfig config,
        EventCapture obs)
    {
        _inner = inner;
        _config = config;
        _obs = obs;
    }

    public OrderAccepted SubmitOrder(string orderId)
    {
        if (_config.SingleInflight && !_state.TryAdd(orderId))
        {
            _obs.Emit(new() { { "ev", "ORDER_REJECTED_DUPLICATE" }, { "order_id", orderId } });
            throw new InvalidOperationException("ORDER_ALREADY_IN_FLIGHT");
        }

        try
        {
            return _inner.SubmitOrder(orderId);
        }
        finally
        {
            _state.Remove(orderId);
        }
    }
}

// ============================================================
// Tests
// ============================================================

[Trait("Level", "cmp_int")]
[Trait("Family", "guard_behavior")]
public sealed class GuardBehaviorTest
{
    [Fact]
    public void RejectsDuplicateInflightRequestAndProviderIsNotCalledTwice()
    {
        // Arrange
        var obs = new EventCapture();
        var provider = new BlockingOrderProvider();
        var guarded = new GuardedOrderPort(
            provider,
            new SubmitOrderGuardConfig(SingleInflight: true),
            obs);

        OrderAccepted? firstResult = null;
        Exception? secondError = null;

        // Act — first call blocks in provider
        var firstCall = new Thread(() =>
        {
            firstResult = guarded.SubmitOrder("ORDER_001");
        });
        firstCall.Start();
        provider.Entered.Wait(TimeSpan.FromSeconds(1))
            .Should().BeTrue("first call must enter provider");

        // Second call while first is in-flight
        try { guarded.SubmitOrder("ORDER_001"); }
        catch (Exception ex) { secondError = ex; }

        // Release the first call
        provider.Release.Set();
        firstCall.Join(TimeSpan.FromSeconds(1));

        // Assert — first call succeeded
        firstResult.Should().NotBeNull();
        firstResult.Should().Be(
            new OrderAccepted("ORDER_001", "accepted"));

        // Assert — duplicate rejected
        secondError.Should().NotBeNull();
        secondError.Should().BeOfType<InvalidOperationException>();
        secondError!.Message.Should().Contain("ORDER_ALREADY_IN_FLIGHT");

        // Assert — provider called exactly once
        provider.Calls.Should().BeEquivalentTo(["ORDER_001"]);

        // Assert — observability captured the rejection
        obs.Events.Should().HaveCount(1);
        obs.Events[0]["ev"].Should().Be("ORDER_REJECTED_DUPLICATE");
        obs.Events[0]["order_id"].Should().Be("ORDER_001");
    }
}

File: eng/docs/agent_skills/python_typing/examples/integration tests/guard_behavior_test.py
from __future__ import annotations

from dataclasses import dataclass, field
from threading import Event, Thread

import pytest

from dictionary.implementation.TL_CORE import ApiError, ErrorCategory, Result


@dataclass(frozen=True, slots=True)
class OrderAccepted:
    order_id: str
    status: str


@dataclass(frozen=True, slots=True)
class SubmitOrderGuardConfig:
    single_inflight: bool = True


@dataclass(slots=True)
class SubmitOrderGuardState:
    inflight_ids: set[str] = field(default_factory=set)


class EventCapture:
    def __init__(self) -> None:
        self.events: list[dict] = []

    def emit(self, event: dict) -> None:
        self.events.append(event)


class OrderPort:
    def submit_order(self, order_id: str) -> Result[OrderAccepted]:
        raise NotImplementedError


class BlockingOrderProvider(OrderPort):
    def __init__(self) -> None:
        self.calls: list[str] = []
        self.entered = Event()
        self.release = Event()

    def submit_order(self, order_id: str) -> Result[OrderAccepted]:
        self.calls.append(order_id)
        self.entered.set()
        self.release.wait(timeout=1.0)
        return Result(
            outcome="ok",
            ok=OrderAccepted(order_id=order_id, status="accepted"),
        )


class GuardedOrderPort(OrderPort):
    def __init__(self, inner: OrderPort, cfg: SubmitOrderGuardConfig, obs: EventCapture) -> None:
        self._inner = inner
        self._cfg = cfg
        self._obs = obs
        self._state = SubmitOrderGuardState()

    def submit_order(self, order_id: str) -> Result[OrderAccepted]:
        if self._cfg.single_inflight and order_id in self._state.inflight_ids:
            self._obs.emit({"ev": "ORDER_REJECTED_DUPLICATE", "order_id": order_id})
            return Result(
                outcome="err",
                err=ApiError(
                    category=ErrorCategory.CONFLICT,
                    code="ORDER_ALREADY_IN_FLIGHT",
                    retryable=True,
                ),
            )

        self._state.inflight_ids.add(order_id)
        try:
            return self._inner.submit_order(order_id)
        finally:
            self._state.inflight_ids.discard(order_id)


@pytest.fixture
def guarded_order_slice() -> tuple[GuardedOrderPort, BlockingOrderProvider, EventCapture]:
    obs = EventCapture()
    provider = BlockingOrderProvider()
    guarded = GuardedOrderPort(inner=provider, cfg=SubmitOrderGuardConfig(), obs=obs)
    return guarded, provider, obs


@pytest.mark.cmp_int
@pytest.mark.guard_behavior
def test_guard_rejects_duplicate_inflight_request_and_provider_is_not_called_twice(
    guarded_order_slice: tuple[GuardedOrderPort, BlockingOrderProvider, EventCapture],
) -> None:
    guarded, provider, obs = guarded_order_slice
    first_results: list[Result[OrderAccepted]] = []

    def run_first_call() -> None:
        first_results.append(guarded.submit_order("ORDER_001"))

    first_call = Thread(target=run_first_call)
    first_call.start()
    assert provider.entered.wait(timeout=1.0)

    duplicate_result = guarded.submit_order("ORDER_001")

    provider.release.set()
    first_call.join(timeout=1.0)

    assert len(first_results) == 1
    assert first_results[0] == Result(
        outcome="ok",
        ok=OrderAccepted(order_id="ORDER_001", status="accepted"),
    )
    assert duplicate_result.outcome == "err"
    assert duplicate_result.ok is None
    assert duplicate_result.err == ApiError(
        category=ErrorCategory.CONFLICT,
        code="ORDER_ALREADY_IN_FLIGHT",
        retryable=True,
    )
    assert provider.calls == ["ORDER_001"]
    assert obs.events == [{"ev": "ORDER_REJECTED_DUPLICATE", "order_id": "ORDER_001"}]
